feat: add `build.getPublicHash` option

aleclarson opened the pull request:
Generate a revision hash for each asset in the public directory.

```ts
import revHash from 'rev-hash'
import fs from 'fs'

export default {
  build: {
    getPublicHash: file => revHash(fs.readFileSync(file)),
  },
}
```

### Features
- Easy HTTP caching of `public/` assets. Cache forever, effortless "invalidation".
- **Internal:** Avoids `checkPublicFile` calls when building, thereby reducing FS calls.

aleclarson committed:
feat: add `build.getPublicHash` option

Generate a revision hash for each asset in the public directory.

yyx990803 commented:
I don't know about this. Assets referenced from imports are already auto-hashed. So this means anything under `assets` are hashed - you just need to configure cache headers for those files.

The whole of `public` files is for those files that **need to preserve exact filenames**. It doesn't make sense to hash them.

aleclarson committed:
fix: ensure parent dirs exist

aleclarson commented:
I've been using `public/` for all of the `.svg`, `.png`, and fonts in my website, since it takes zero configuration for short import paths (eg: `/svg/foo.svg`), and I prefer keeping static assets out of the `src` directory (without having both a `public` directory *and* an `assets` directory).

The alternatives are:
- use relative import paths (less elegant, harder to maintain)
- use `paths` feature in tsconfig.json and install [`vite-tsconfig-paths`](https://github.com/aleclarson/vite-tsconfig-paths)
- custom plugin that rewrites `/svg/*` imports to `config.root + "/src/svg/[name]"`

aleclarson committed:
refactor: remove assetsCopied hook + more

- ensure JS bundle references to public/ assets are rewritten
- add `publicUrls` property to ResolvedConfig
- copy public assets before HTML transforms are applied

aleclarson commented:
@knightly-bot build this

yyx990803 commented:
I would prefer this to be a 3rd party plugin to not break the originally intended usage of `public`.

aleclarson commented:
I tried that approach first, but it's not currently possible. If we had a hook for rewriting public filenames (eg: `resolvePublicId`), it would be pretty easy, I think.

aleclarson commented:
Removed the `assetsCopied` hook, in favor of `publicUrls` object added to `ResolvedConfig`.

aleclarson commented:
Closed in favor of #1647 and #1648
